Sources/AuthFoundation/User Management/Credential.swift

```diff
@@ -301,55 +337,127 @@ public final class Credential: Equatable, OAuth2ClientDelegate {
     
     // MARK: Private properties
     static let coordinator = CredentialCoordinatorImpl()
-    weak var coordinator: CredentialCoordinator?
+    nonisolated(unsafe) weak var coordinator: CredentialCoordinator?
 
+    @CredentialActor
     static func resetToDefault() {
         coordinator.resetToDefault()
+        _CredentialAutomaticRefreshTimeIntervalToNanoseconds = 1_000_000_000
     }
 
+    nonisolated(unsafe) private var _token: Token
     let userInfoAction = CoalescedResult<UserInfo>(taskName: "UserInfo")
     let lock = Lock()
 
-    private lazy var _metadata: Token.Metadata = {
-        if let metadata = try? coordinator?.tokenStorage.metadata(for: token.id) {
-            return metadata
+    nonisolated(unsafe) private var _metadata: Token.Metadata?
+    var metadata: Token.Metadata {
+        get {
+            lock.withLock {
+                if let metadata = _metadata {
+                    return metadata
+                }
+
+                let result: Token.Metadata
+
+                let id = id
+                if let coordinator,
+                   let metadata = withAsyncGroup({
+                       try? await coordinator.tokenStorage.metadata(for: id)
+                   })
+                {
+                    result = metadata
+                } else {
+                    result = Token.Metadata(id: id)
+                }
+
+                _metadata = result
+                return result
+            }
         }
-        
-        return Token.Metadata(id: id)
-    }()
-    
-    internal private(set) var automaticRefreshTimer: DispatchSourceTimer?
+        set {
+            lock.withLock {
+                _metadata = newValue
+            }
+        }
+    }
+
+    nonisolated(unsafe) private var _automaticRefresh: Bool = false
+    nonisolated(unsafe) private(set) var _automaticRefreshTimer: DispatchSourceTimer?
+    nonisolated(unsafe) private var _automaticRefreshTask: Task<Void, Never>?
+
     private func startAutomaticRefresh() {
-        guard let timerSource = createAutomaticRefreshTimer() else { return }
+        guard let expiresAt = _token.expiresAt
+        else {
+            return
+        }
+        let graceInterval = Credential.refreshGraceInterval
+        let timeOffset = max(0.0, expiresAt.timeIntervalSinceNow - Date.nowCoordinated.timeIntervalSinceNow - graceInterval)
+        let repeatInterval = min(_token.expiresIn - graceInterval, _token.expiresIn)
+
+        _automaticRefreshTask = Task(priority: .medium) {
```

Comment:
What if task is already created and active?

Reply:
`start` and `stopAutomaticRefresh` are only invoked from the `automaticRefresh` property's setter.  But even so, in the event a future code change is made which breaks this, once the old Task's cancellation handler is overwritten, it will lose its retention count and will be cancelled & destroyed.